### ISSUE TITLE
bug in BuildTestCommand, only retrieving first line

### ISSUE THREAD

**Issue by shahzebsiddiqui** [buildtesters/buildtest#287]
Here is an example of BuildTestCommand, its retrieving the first line of output which is wrong.
This affects the content of output and error stream written to file.
```
>>> from buildtest.utils.command import BuildTestCommand
>>> cmd = BuildTestCommand("python .buildtest/python-circle/circle_area.py")
>>> out,err = cmd.execute()
>>> out
['Circle Radius  2\n']

>>> os.system("python .buildtest/python-circle/circle_area.py")
Circle Radius  2
Area of circle  12.566370614359172
Block
0
```

**Comment by vsoch** [comment]
This was definitely working when I developed it - if you are able to track what changed (supporting functions too) that likely will find the culprit.

**Comment by shahzebsiddiqui** [comment]
According to https://github.com/buildtesters/buildtest/tree/devel/buildtest/utils i dont see any commits on the file that would have changed from your PR https://github.com/buildtesters/buildtest/pull/271 

Just did another test by using a list as pose to string and this seems like its same result

```
>>> from buildtest.utils.command import BuildTestCommand
>>> cmd = BuildTestCommand(["python",".buildtest/python-circle/circle_area.py"])
>>> out,err = cmd.execute()
>>> out
['Circle Radius  2\n']
```

It looks like you had it working at one point in this comment https://github.com/buildtesters/buildtest/pull/271#issue-402411911  but now it's not.

**Comment by vsoch** [comment]
It's still working for me locally for that same case:

```
from buildtest.utils.command import *                                                                                                                                    
cmd = BuildTestCommand(['/bin/bash', 'script.sh'])                                                                                                                       
cmd.execute()                                                                                                                                                            
       
Out[4]: (['LOOP 1\n', 'LOOP 2\n', 'LOOP 3\n', 'LOOP 4\n', 'LOOP 5\n', 'LOOP 6\n'], [])

cmd.out                                                                                                                                                                  
['LOOP 1\n', 'LOOP 2\n', 'LOOP 3\n', 'LOOP 4\n', 'LOOP 5\n', 'LOOP 6\n']
cmd.err                                                                                                                                                                  
[]
```
let me give your case a look.

**Comment by vsoch** [comment]
Share the content of the python script here please?

**Comment by shahzebsiddiqui** [comment]
You can build this by running ``buildtest build -b examples/script/python-circle.yml`` and check the generated file in your ``.buildtest``. Mine is the following

```
(buildtest-framework) ssi29@ag-mxg-hulk090> cat .buildtest/python-circle/circle_area.py
#!/u/users/ssi29/.local/share/virtualenvs/buildtest-framework-LDt8wyRf/bin/python
import os
import math

radius = 2
area = math.pi * radius * radius
print("Circle Radius ", radius)
print("Area of circle ", area)
if True:
    print("Block")

```

**Comment by vsoch** [comment]
I can't reproduce your error:

```
from buildtest.utils.command import *                                                                                                                                    
cmd = BuildTestCommand("python script.py")                                                                                                                               

cmd.execute()                                                                                                                                                            
(['Circle Radius  2\n', 'Area of circle  12.566370614359172\n'], [])
```
My head is at:

```
commit 2889a2934615364d4b6b03792677f1226e4fcea3 (HEAD -> fix/command-out-err, origin/fix/command-out-err)
Author: vsoch <vsochat@stanford.edu>
Date:   Sun Apr 12 19:01:04 2020 -0600

    fixing issues with returncode for BuildTestCommand
    
    Signed-off-by: vsoch <vsochat@stanford.edu>
```
so I suspect that some following change that you made is leading to the issue.

**Comment by shahzebsiddiqui** [comment]
Okay so i recreated my python environment (pipenv) and used ``devel`` as source of test by running ``buildtest build -b examples/script/python-circle.yml`` as a test since this gives an output of more than one line.

I see only one line printed

```
(buildtest) ssi29@ag-mxg-hulk090> cat .buildtest/python-circle/run/circle_area_04-24-2020-11-49.out
Circle Radius  2
```

Running the test manually i still see 3 lines

```
(buildtest) ssi29@ag-mxg-hulk090> python circle_area.py
Circle Radius  2
Area of circle  12.566370614359172
Block
```

Now testing this with ``BuildTestCommand`` on same run. There is an issue that is present on ``devel``.

```
>>> from buildtest.utils.command import *
>>> cmd = BuildTestCommand("python circle_area.py")
>>> cmd.execute()
(['Circle Radius  2\n'], [])
```

I am getting same results when using the commit HEAD when you made change. Can you confirm by building any tests from ``tutorials`` repo and see if output is generated properly. There are some tests that are meant to generate more than one line. For example this test https://github.com/buildtesters/tutorials/blob/master/apps/openmp/helloworld/omp_hello.c.yml is meant to print several lines of output.

**Comment by vsoch** [comment]
I don't think me testing more examples (and showing that it works) will make any progress - I think this must be a system issue? Did these commands ever work for you (or was it just me?) Did you test when I showed the example in the GitHub comment? I can't reproduce your error, so I think I would suggest interactively (in ipython or similar) testing issuing different commands on your system to see if you can get something to work. if you are working in a headless environment, try running it on a local machine (and one more similar to mine) to see if you can reproduce working.

In debugging, it's easier to help when someone can reproduce the issue, and then work on that. I'm not able to reproduce it (but you are) so we should take the opposite approach now - you should try to reproduce the working one by changing your environment/host and then figuring out the difference s between those two. I have no way I can offer to help since I can't reproduce anything.

**Comment by shahzebsiddiqui** [comment]
Let me try testing on another cluster. I don't have a local machine running Linux. I seem to recall this was working for me but all of a sudden its not.  I figured it may be issue with my python environment but i have rebuild my pipenv env so i don't think that is the issue.

**Comment by shahzebsiddiqui** [comment]
Okay i so i am getting mixed results, when running ``buildtest build -b apps/openmp/helloworld/`` I get the output file generated properly.

```
(buildtest-framework) login4.stampede2(1036)$ cat .buildtest/omp_hello.c/run/hello_world_04-24-2020-15-17.out
Hello World from thread = 0

Number of threads = 2

Hello World from thread = 1

```

Now running something from the client i am not getting the right results.
```
>>> from buildtest.utils.command import *
>>> cmd = BuildTestCommand("lscpu")
>>> cmd.execute()
(['Architecture:          x86_64\n'], [])
```
The output of ``lscpu`` should me more than just one line.


Now running the same openmp code on another machine i only get one line. It is certainly system dependent

```
(buildtest) ssi29@ag-mxg-hulk090> cat .buildtest/omp_hello.c/run/hello_world_04-24-2020-16-21.out
Hello World from thread = 9
```

I think the client ``BuildTestCommand`` should work when using interactively, so having this not work from two different machine does not sound promising. Is it possible to rethink the way BuildTestCommand is processing each line and polling for stdout. If ``BuildTestCommand`` is not writing output correctly, we can't really do proper regex check since its dependent on output stream to be written properly.

**Comment by vsoch** [comment]
I think the right step to take here is figure out why there are differences. We can't really edit the code / refactor until that is clear, because we need to know what error and output streams are being used, etc.

**Comment by shahzebsiddiqui** [comment]
So i see the issue is that returncode is ``None``  when running ``process.poll`` in line https://github.com/buildtesters/buildtest/blob/devel/buildtest/utils/command.py#L60 and this means child process didn't terminate properly https://docs.python.org/3.8/library/subprocess.html#subprocess.Popen.poll 

This means it never reaches the while loop and terminates with current line. I was troubleshooting this to print the return code right after the poll and before the while loop.

```
>>> from buildtest.utils.command import *
>>> cmd = BuildTestCommand("lscpu")
>>> cmd.execute()
Executing command: ['/usr/bin/lscpu']
Return Code: None
(['Architecture:          x86_64\n'], [])
```

I tested this manually via ``subprocess`` and returncode is 0. 

```
>>> import subprocess
>>> process = subprocess.Popen(['/usr/bin/lscpu'],stderr=subprocess.PIPE,stdout=subprocess.PIPE)
>>> returncode = process.poll()
>>> returncode
0
```

I don't particularly understand the reasoning why we were reading one line at a time. Is there a reason why we can't use builtin methods from subprocess library to get the output to simplify this code.

**Comment by vsoch** [comment]
subprocess.poll is expected to return None until execution is done. It’s the way to stream the out and err line by line instead of needing to wait a long time. Try something with a long time to execute or sleep command to see the difference.

I don’t think the right way to go about this is just changing the code without knowing what’s going on. You need to figure out why your environment is not returning stderr and stdout like every other environment and then decide how to change the code.

**Comment by vsoch** [comment]
I can’t reproduce your error, so I don’t know how I’d work on this sorry.

**Comment by shahzebsiddiqui** [comment]
Since you re-implemented this feature i am assigning this to you to figure out the problem. There is clearly some problem on ``devel`` and there seems to be no change to file since PR https://github.com/buildtesters/buildtest/pull/271 

Me troubleshooting is not going to help when i don't understand the reasoning behind the implementation, previously it was much simpler granted that it was using ``shell=True``.

**Comment by shahzebsiddiqui** [comment]
> I can’t reproduce your error, so I don’t know how I’d work on this sorry.

Please test by running output of any Linux command such as ``lscpu``. It would help to know your system environment including operating system, python environment using devel as the HEAD. If there are differences between devel and your PR https://github.com/buildtesters/buildtest/commit/c343526ef53cbe1e15e89f9fd1c13d21e10fc6b4 that would help narrow the issue. 

It would help if you can test on two different system (cluster and local machine) and i can reproduce the same behavior. I was tempted to copy paste the whole code block but didn't see any difference so it didn't help.

**Comment by vsoch** [comment]
Fixed with #289
